eclair-core/src/main/scala/fr/acinq/eclair/router/BalanceEstimate.scala

```diff
@@ -0,0 +1,263 @@
+   *     0             low     high                       capacity
+   */
+
+  /**
+   * We model the decay with a half-life H: every H units of time, our confidence decreases by half and our estimated
+   * probability distribution gets closer to the baseline uniform distribution of balances between 0 and totalCapacity.
+   *
+   * @param amount                the amount that we knew we could send or not send at time t
+   * @param successProbabilityAtT probability that we could relay amount at time t (usually 0 or 1)
+   * @param t                     time at which we knew if we could or couldn't send amount
+   * @return                      the probability that we can send amount now
+   */
+  private def decay(amount: MilliSatoshi, successProbabilityAtT: Double, t: TimestampSecond): Double = {
+    val decayRatio = 1 / math.pow(2, (TimestampSecond.now() - t) / halfLife)
+    val baseline = 1 - amount.toLong.toDouble / toMilliSatoshi(totalCapacity).toLong.toDouble
+    baseline * (1 - decayRatio) + successProbabilityAtT * decayRatio
+  }
+
+  def otherSide: BalanceEstimate = BalanceEstimate(totalCapacity - high, highTimestamp, totalCapacity - low, lowTimestamp, totalCapacity, halfLife)
+
+  def couldNotSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    if (amount < high) {
+      if (amount > low) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        // the balance is actually below `low`, we discard our lower bound
+        copy(low = MilliSatoshi(0), lowTimestamp = timestamp, high = amount, highTimestamp = timestamp)
+      }
+    } else {
+      /* We already expected not to be able to relay that amount as it above our upper bound. However if the upper bound
+       * was old enough that replacing it with the current amount decreases the success probability for `high`, then we
+       * replace it.
+       */
+      val pLow = decay(low, 1, lowTimestamp)
+      val pHigh = decay(high, 0, highTimestamp)
+      val x = low + (high - low) * (pLow / (pLow - pHigh))
+      if (amount <= x) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        copy()
+      }
+    }
+
+  def couldSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    otherSide.couldNotSend(totalCapacity - amount, timestamp).otherSide
+
+  def didSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate = {
+    val newLow = (low - amount) max MilliSatoshi(0)
+    val newHigh = (high - amount) max MilliSatoshi(0)
+    /* We could shift everything left by amount without changing the timestamps (a) but we may get more information by
+     * ignoring the old high (b) if if has decayed too much. We try both and choose the one that gives the lowest
+     * probability for high.
+     */
+    val a = copy(low = newLow, high = newHigh)
+    val b = copy(low = newLow, high = (totalCapacity - amount) max MilliSatoshi(0), highTimestamp = timestamp)
```

Comment:
Shouldn't this one also update the low timstamp?

Reply:
No, sending some amount does not make the lower bound apply to the current timestamp.

Reply:
I understand, I think you're right.